File: frontend/src/engine/finance/importrules.test.ts
// Mirror of backend/finance/import_rules_test.go: item kinds and billing
// months from a statement, the confirm paths' kind/currency guards, one link
// per expense, reopening confirmed items, and fixed-expense bills.
import { beforeEach, describe, expect, it } from 'vitest'
import { createTestDb } from '@/engine/testing/db'
import { createFinanceService } from '@/engine/finance/service'
import { createSession } from '@/engine/users/service'
import { namesMatch } from '@/engine/finance/fixedmatch'
import { blankStatement, statementLine as line } from '@/engine/testing/cardStatement'
import type { FinanceServiceContract, ImportCandidate, ImportItemView } from '@/services/contract'

let finance: FinanceServiceContract

beforeEach(async () => {
  const db = await createTestDb()
  finance = createFinanceService(db, createSession(db))
})

const candidate = (c: Partial<ImportCandidate>): ImportCandidate => ({
  date: '',
  description: '',
  amount: '',
  currency: '',
  cardLastDigits: '',
  account: '',
  reference: '',
  installmentsTotal: 0,
  hint: '',
  ...c,
})

async function pending(): Promise<ImportItemView[]> {
  const r = await finance.ListImportItems('pendiente')
  expect(r.error).toBeUndefined()
  return r.data ?? []
}

async function stageOne(c: Partial<ImportCandidate>): Promise<ImportItemView> {
  const r = await finance.StageImport({ source: 'pdf_account', issuer: 'Itau', items: [candidate(c)] })
  expect(r.error).toBeUndefined()
  const it = (await pending()).find((i) => i.description === c.description && i.date === c.date)
  if (!it) throw new Error(`staged item ${c.description ?? ''} not pending`)
  return it
}

describe('statement lines', () => {
  it('fijan el tipo del movimiento y el mes de facturación', async () => {
    const r = await finance.ImportCardStatement({
      ...blankStatement,
      issuer: 'itau', kind: 'nacional', currency: 'CLP', cardLastDigits: '4321',
      statementDate: '2026-08-25', periodTo: '2026-08-25', totalBilled: '0',
      lines: [
        line({ section: 'compra', operationDate: '2026-08-12', reference: 'a', description: 'TAXI VIAJE',
          operationAmount: '2340', installmentAmount: '2340' }),
        line({ section: 'compra', operationDate: '2026-06-20', reference: 'b', description: 'TIENDA DOS',
          operationAmount: '60000', installmentNumber: 3, installmentsTotal: 6, installmentAmount: '10000' }),
        line({ section: 'cargo', operationDate: '2026-08-25', reference: 'c', description: 'COMISION MENSUAL',
          operationAmount: '20000', installmentAmount: '20000' }),
        line({ section: 'compra', operationDate: '2026-08-14', reference: 'd', description: 'ANULACION COMPRA',
          operationAmount: '-15990', installmentAmount: '-15990' }),
      ],
    })
    expect(r.error).toBeUndefined()
    const byDesc = new Map((await pending()).map((i) => [i.description, i]))
    expect(byDesc.get('TAXI VIAJE')).toMatchObject({ kind: 'gasto', firstPeriod: '2026-08' })
    expect(byDesc.get('TIENDA DOS')).toMatchObject({ kind: 'gasto', firstPeriod: '2026-06' })
    expect(byDesc.get('COMISION MENSUAL')).toMatchObject({ kind: 'gasto', firstPeriod: '2026-08' })
    expect(byDesc.get('ANULACION COMPRA')).toMatchObject({ kind: 'abono', amount: '15990' })
  })
})

describe('confirm paths', () => {
  it('exigen el tipo correcto y pesos para montos en otra moneda', async () => {
    const credit = await stageOne({ date: '2026-08-17', description: 'ABONO CANJE', amount: '3145', kind: 'abono' })
    const charge = await stageOne({ date: '2026-08-12', description: 'TAXI', amount: '5130' })
    const usd = await stageOne({ date: '2026-08-02', description: 'ANTHROPIC CLAUDE SUB', amount: '119', currency: 'USD' })
    const confirm = (id: number, amount: string) =>
      finance.ConfirmImportItem(id, '2026-08-02', 'x', '', '', null, 'unico', amount, 1, '')

    expect((await confirm(credit.id, '3145')).error?.code).toBe('VALIDATION_ERROR')
    expect((await finance.LinkImportItem(credit.id, 1)).error?.code).toBe('VALIDATION_ERROR')
    expect((await finance.ConfirmImportItemAsIncome(charge.id, '2026-08', 'x', '5130')).error?.code).toBe(
      'VALIDATION_ERROR',
    )
    expect((await confirm(usd.id, '119')).error?.code).toBe('VALIDATION_ERROR')
    expect((await confirm(usd.id, '112948.5')).error?.code).toBe('VALIDATION_ERROR')
    expect((await confirm(usd.id, '112948')).error).toBeUndefined()
    expect((await finance.ConfirmImportItemAsIncome(credit.id, '2026-08', 'Canje', '3145')).error).toBeUndefined()
  })

  it('un gasto acepta un solo movimiento enlazado', async () => {
    const ex = await finance.CreateExpense('2026-08-12', 'Taxi', '', '', null, 'unico', '5130', 1)
    const first = await stageOne({ date: '2026-08-12', description: 'TAXI UNO', amount: '5130' })
    const second = await stageOne({ date: '2026-08-12', description: 'TAXI DOS', amount: '5130' })
    expect((await finance.LinkImportItem(first.id, ex.data!.id)).error).toBeUndefined()
    expect((await finance.LinkImportItem(second.id, ex.data!.id)).error?.code).toBe('CONFLICT')
  })

  it('un confirmado se puede reabrir cuando su gasto va a la papelera', async () => {
    const it0 = await stageOne({ date: '2026-08-12', description: 'TAXI', amount: '5130' })
    const r = await finance.ConfirmImportItem(it0.id, '2026-08-12', 'Taxi', '', '', null, 'unico', '5130', 1, '')
    expect((await finance.RestoreImportItem(it0.id)).error?.code).toBe('CONFLICT')
    expect((await finance.ListImportItems('confirmado')).data?.[0]?.reopenable).toBe(false)

    await finance.DeleteExpense(r.data!.id)
    expect((await finance.ListImportItems('confirmado')).data?.[0]?.reopenable).toBe(true)
    expect((await finance.RestoreImportItem(it0.id)).error).toBeUndefined()
    const back = await pending()
    expect(back).toHaveLength(1)
    expect(back[0]?.expenseId).toBeNull()
  })
})

describe('fixed-expense bills', () => {
  it('sugiere el gasto fijo y al enlazar marca el mes pagado con el monto real', async () => {
    const fe = await finance.CreateFixedExpense('Plan Entel', 'Servicios', null, '2026-07', '17000')
    const bill = await stageOne({ date: '2026-08-05', description: 'ENTEL PCS PAGO ENSANTIAGO C', amount: '16990' })
    expect(bill.suggestedFixedId).toBe(fe.data!.id)
    expect(bill.suggestedFixedPeriod).toBe('2026-08')
    for (const c of [
      { date: '2026-08-06', description: 'TRANSFERENCIA A JUAN', amount: '17000' },
      { date: '2026-08-07', description: 'ENTEL TIENDA EQUIPO', amount: '250000' },
      { date: '2026-06-05', description: 'ENTEL PCS PAGO JUNIO', amount: '17000' },
    ]) {
      expect((await stageOne(c)).suggestedFixedId, c.description).toBeNull()
    }

    expect((await finance.LinkImportItemToFixed(bill.id, fe.data!.id, '2026-08')).error).toBeUndefined()
    for (const [period, amount, status] of [
      ['2026-08', '16990', 'pagado'],
      ['2026-09', '17000', 'pendiente'],
    ] as const) {
      const mov = (await finance.MonthlySummary(period)).data?.movimientos.find((m) => m.fixedId === fe.data!.id)
      expect(mov, period).toMatchObject({ amount, status })
    }

    const again = await stageOne({ date: '2026-08-20', description: 'ENTEL PCS PAGO OTRO', amount: '16990' })
    expect(again.suggestedFixedId).toBeNull()
    expect((await finance.LinkImportItemToFixed(again.id, fe.data!.id, '2026-08')).error?.code).toBe('CONFLICT')
    expect((await finance.LinkImportItemToFixed(again.id, fe.data!.id, '2026-06')).error?.code).toBe(
      'VALIDATION_ERROR',
    )
  })

  it('namesMatch compara palabras significativas', () => {
    expect(namesMatch('Plan Entel', 'ENTEL PCS PAGO ENSANTIAGO C')).toBe(true)
    expect(namesMatch('Claude', 'ANTHROPIC* CLAUDE SUB')).toBe(true)
    expect(namesMatch('Proseguro', 'PROSEGUR ACTIVA')).toBe(true)
    expect(namesMatch('Netflix', 'APPLE.COM/BILL')).toBe(false)
    expect(namesMatch('Luz', 'ENEL DISTRIBUCION')).toBe(false)
  })
})

describe('international payments', () => {
  it('un pago ya descartado igual concilia y enseña la tasa', async () => {
    const payment = await stageOne({
      date: '2026-07-31', description: 'PAGO DEUDA INTER. TC CTA CLP', amount: '94900', hint: 'card_payment',
    })
    await finance.DiscardImportItem(payment.id)
    const r = await finance.ImportCardStatement({
      ...blankStatement,
      issuer: 'itau', kind: 'internacional', currency: 'USD', cardLastDigits: '4321',
      statementDate: '2026-08-25', periodTo: '2026-08-25', totalBilled: '20',
      lines: [
        line({ section: 'pago', operationDate: '2026-07-30', reference: '3007', description: 'MONTO CANCELADO',
          installmentAmount: '-100', originAmount: '-100' }),
        line({ section: 'compra', operationDate: '2026-08-02', reference: '0308 99', description: 'SERVICIO WEB SUB',
          installmentAmount: '20', originAmount: '20' }),
      ],
    })
    expect(r.data?.paymentsMatched).toBe(1)
    expect((await pending()).find((i) => i.description === 'SERVICIO WEB SUB')?.suggestedAmountClp).toBe('18980')
  })
})

File: backend/finance/import_rules_test.go
package finance

import (
	"testing"

	"github.com/gastonlarap-a11y/app-finance/backend/shared"
	"github.com/gastonlarap-a11y/app-finance/backend/shared/types"
)

func dec(t *testing.T, s string) types.Decimal {
	t.Helper()
	d, err := types.New(s)
	if err != nil {
		t.Fatalf("decimal %q: %v", s, err)
	}
	return d
}

func TestLineCandidateKindAndBillingMonth(t *testing.T) {
	st := &CardStatement{Kind: StatementNational, Currency: "CLP", Period: "2026-08"}
	for _, tc := range []struct {
		name            string
		line            CardStatementLine
		wantKind        string
		wantAmount      string
		wantFirstPeriod string
	}{
		{"one-payment purchase is billed in the statement month",
			CardStatementLine{Section: LinePurchase, OperationAmount: dec(t, "2340"), InstallmentAmount: dec(t, "2340"), InstallmentNumber: 1, InstallmentsTotal: 1},
			ImportKindExpense, "2340", "2026-08"},
		{"cuota 3/6 started two months before",
			CardStatementLine{Section: LinePurchase, OperationAmount: dec(t, "60000"), InstallmentAmount: dec(t, "10000"), InstallmentNumber: 3, InstallmentsTotal: 6},
			ImportKindExpense, "60000", "2026-06"},
		{"a fee is billed in the statement month",
			CardStatementLine{Section: LineCharge, OperationAmount: dec(t, "20000"), InstallmentAmount: dec(t, "20000"), InstallmentNumber: 1, InstallmentsTotal: 1},
			ImportKindExpense, "20000", "2026-08"},
		{"a reversed purchase is money back",
			CardStatementLine{Section: LinePurchase, OperationAmount: dec(t, "-15990"), InstallmentAmount: dec(t, "-15990"), InstallmentNumber: 1, InstallmentsTotal: 1},
			ImportKindCredit, "15990", ""},
		{"a refunded fee is money back",
			CardStatementLine{Section: LineCharge, OperationAmount: dec(t, "-20000"), InstallmentAmount: dec(t, "-20000"), InstallmentNumber: 1, InstallmentsTotal: 1},
			ImportKindCredit, "20000", ""},
		{"a credit stays a credit",
			CardStatementLine{Section: LineCredit, OperationAmount: dec(t, "-10000"), InstallmentAmount: dec(t, "-10000"), InstallmentNumber: 1, InstallmentsTotal: 1},
			ImportKindCredit, "10000", ""},
	} {
		t.Run(tc.name, func(t *testing.T) {
			c := lineCandidate(st, &tc.line)
			kind := c.Kind
			if kind == "" {
				kind = ImportKindExpense
			}
			if kind != tc.wantKind || c.Amount != tc.wantAmount || c.FirstPeriod != tc.wantFirstPeriod {
				t.Fatalf("candidate = kind %q amount %q first %q, want %q %q %q",
					kind, c.Amount, c.FirstPeriod, tc.wantKind, tc.wantAmount, tc.wantFirstPeriod)
			}
		})
	}
}

// stageOne stages a single pdf_account candidate and returns its pending view.
func stageOne(t *testing.T, s *FinanceService, c ImportCandidate) ImportItemView {
	t.Helper()
	mustStage(t, s, pdfBatch(c))
	for _, it := range mustList(t, s, ImportPendiente) {
		if it.Description == c.Description && it.Date == c.Date {
			return it
		}
	}
	t.Fatalf("staged item %q not pending", c.Description)
	return ImportItemView{}
}

func TestConfirmPathsRequireKindAndPesos(t *testing.T) {
	ctx := t.Context()
	s := newTestService(t)
	credit := stageOne(t, s, ImportCandidate{Date: "2026-08-17", Description: "ABONO CANJE", Amount: "3145", Kind: ImportKindCredit})
	charge := stageOne(t, s, ImportCandidate{Date: "2026-08-12", Description: "TAXI", Amount: "5130"})
	usd := stageOne(t, s, ImportCandidate{Date: "2026-08-02", Description: "ANTHROPIC CLAUDE SUB", Amount: "119", Currency: "USD"})

	confirm := func(id int64, amount string) *shared.AppError {
		return s.ConfirmImportItem(ctx, id, "2026-08-02", "x", "", "", nil, KindUnico, amount, 1, "").Error
	}
	wantCode(t, "credit confirmed as expense", confirm(credit.ID, "3145"), shared.ErrValidation)
	wantCode(t, "credit linked to an expense", s.LinkImportItem(ctx, credit.ID, 1).Error, shared.ErrValidation)
	wantCode(t, "charge confirmed as income",
		s.ConfirmImportItemAsIncome(ctx, charge.ID, "2026-08", "x", "5130").Error, shared.ErrValidation)
	wantCode(t, "USD amount copied as pesos", confirm(usd.ID, "119"), shared.ErrValidation)
	wantCode(t, "USD amount with decimals", confirm(usd.ID, "112948.5"), shared.ErrValidation)

	mustOK(t, "USD converted to pesos", confirm(usd.ID, "112948"))
	mustOK(t, "credit as income", s.ConfirmImportItemAsIncome(ctx, credit.ID, "2026-08", "Canje", "3145").Error)
}

func TestLinkImportItemOncePerExpense(t *testing.T) {
	ctx := t.Context()
	s := newTestService(t)
	ex := s.CreateExpense(ctx, "2026-08-12", "Taxi", "", "", nil, KindUnico, "5130", 1)
	mustOK(t, "CreateExpense", ex.Error)
	first := stageOne(t, s, ImportCandidate{Date: "2026-08-12", Description: "TAXI UNO", Amount: "5130"})
	second := stageOne(t, s, ImportCandidate{Date: "2026-08-12", Description: "TAXI DOS", Amount: "5130"})

	mustOK(t, "first link", s.LinkImportItem(ctx, first.ID, ex.Data.ID).Error)
	wantCode(t, "second link to the same expense", s.LinkImportItem(ctx, second.ID, ex.Data.ID).Error, shared.ErrConflict)
}

func TestReopenConfirmedItemAfterTrash(t *testing.T) {
	ctx := t.Context()
	s := newTestService(t)
	it := stageOne(t, s, ImportCandidate{Date: "2026-08-12", Description: "TAXI", Amount: "5130"})
	r := s.ConfirmImportItem(ctx, it.ID, "2026-08-12", "Taxi", "", "", nil, KindUnico, "5130", 1, "")
	mustOK(t, "ConfirmImportItem", r.Error)

	wantCode(t, "reopen while the expense is live", s.RestoreImportItem(ctx, it.ID).Error, shared.ErrConflict)
	if got := mustList(t, s, ImportConfirmado); len(got) != 1 || got[0].Reopenable {
		t.Fatalf("confirmed = %+v, want one not reopenable", got)
	}

	mustOK(t, "DeleteExpense", s.DeleteExpense(ctx, r.Data.ID).Error)
	if got := mustList(t, s, ImportConfirmado); len(got) != 1 || !got[0].Reopenable {
		t.Fatalf("confirmed after trashing its expense = %+v, want reopenable", got)
	}
	mustOK(t, "RestoreImportItem", s.RestoreImportItem(ctx, it.ID).Error)
	pending := mustList(t, s, ImportPendiente)
	if len(pending) != 1 || pending[0].ExpenseID != nil {
		t.Fatalf("pending after reopen = %+v, want the item back without its old link", pending)
	}
}

func TestFixedExpenseSuggestionAndLink(t *testing.T) {
	ctx := t.Context()
	s := newTestService(t)
	fe := s.CreateFixedExpense(ctx, "Plan Entel", "Servicios", nil, "2026-07", "17000")
	mustOK(t, "CreateFixedExpense", fe.Error)

	bill := stageOne(t, s, ImportCandidate{Date: "2026-08-05", Description: "ENTEL PCS PAGO ENSANTIAGO C", Amount: "16990"})
	if bill.SuggestedFixedID == nil || *bill.SuggestedFixedID != fe.Data.ID || bill.SuggestedFixedPeriod != "2026-08" {
		t.Fatalf("suggestion = %v %q, want Plan Entel for 2026-08", bill.SuggestedFixedID, bill.SuggestedFixedPeriod)
	}
	for _, c := range []ImportCandidate{
		{Date: "2026-08-06", Description: "TRANSFERENCIA A JUAN", Amount: "17000"}, // amount only
		{Date: "2026-08-07", Description: "ENTEL TIENDA EQUIPO", Amount: "250000"}, // name only
		{Date: "2026-06-05", Description: "ENTEL PCS PAGO JUNIO", Amount: "17000"}, // before it starts
	} {
		if v := stageOne(t, s, c); v.SuggestedFixedID != nil {
			t.Fatalf("%q suggested %s, want no suggestion", c.Description, v.SuggestedFixedDescription)
		}
	}

	mustOK(t, "LinkImportItemToFixed", s.LinkImportItemToFixed(ctx, bill.ID, fe.Data.ID, "2026-08").Error)

	// August is paid at the bank's real amount; September keeps the plan.
	for _, tc := range []struct{ period, amount, status string }{
		{"2026-08", "16990", StatusPagado},
		{"2026-09", "17000", StatusPendiente},
	} {
		sum := s.MonthlySummary(ctx, tc.period)
		mustOK(t, "MonthlySummary", sum.Error)
		var found bool
		for _, m := range sum.Data.Movimientos {
			if m.FixedID != nil && *m.FixedID == fe.Data.ID {
				found = true
				if m.Amount.String() != tc.amount || m.Status != tc.status {
					t.Fatalf("%s fixed movement = %s %s, want %s %s", tc.period, m.Amount, m.Status, tc.amount, tc.status)
				}
			}
		}
		if !found {
			t.Fatalf("%s has no fixed movement", tc.period)
		}
	}

	again := stageOne(t, s, ImportCandidate{Date: "2026-08-20", Description: "ENTEL PCS PAGO OTRO", Amount: "16990"})
	if again.SuggestedFixedID != nil {
		t.Fatalf("a paid month is still suggested: %+v", again)
	}
	wantCode(t, "second link to the same month",
		s.LinkImportItemToFixed(ctx, again.ID, fe.Data.ID, "2026-08").Error, shared.ErrConflict)
	wantCode(t, "link outside the fixed expense's months",
		s.LinkImportItemToFixed(ctx, again.ID, fe.Data.ID, "2026-06").Error, shared.ErrValidation)
}

func TestNamesMatch(t *testing.T) {
	for _, tc := range []struct {
		fixed, descriptor string
		want              bool
	}{
		{"Plan Entel", "ENTEL PCS PAGO ENSANTIAGO C", true},
		{"Claude", "ANTHROPIC* CLAUDE SUB", true},
		{"Proseguro", "PROSEGUR ACTIVA", true},
		{"Netflix", "APPLE.COM/BILL", false},
		{"Luz", "ENEL DISTRIBUCION", false}, // too short to match anything
	} {
		if got := namesMatch(tc.fixed, tc.descriptor); got != tc.want {
			t.Errorf("namesMatch(%q, %q) = %v, want %v", tc.fixed, tc.descriptor, got, tc.want)
		}
	}
}

// A card payment the user already discarded (it is not an expense) still
// reconciles with the statement and teaches the USD rate.
func TestDiscardedPaymentTeachesTheRate(t *testing.T) {
	ctx := t.Context()
	s := newTestService(t)
	payment := stageOne(t, s, ImportCandidate{Date: "2026-07-31", Description: "PAGO DEUDA INTER. TC CTA CLP",
		Amount: "94900", Hint: HintCardPayment})
	mustOK(t, "DiscardImportItem", s.DiscardImportItem(ctx, payment.ID).Error)

	got := importStatement(t, s, internationalStatement()) // pays US$100 on 2026-07-30
	if got.PaymentsMatched != 1 {
		t.Fatalf("PaymentsMatched = %d, want the discarded payment reconciled", got.PaymentsMatched)
	}
	usd := pendingByDescription(t, s)["SERVICIO WEB SUB"]
	if usd.SuggestedAmountClp != "18980" { // US$20 × 949
		t.Fatalf("SuggestedAmountClp = %q, want 18980", usd.SuggestedAmountClp)
	}
}
